StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1130UseLambdaSyntax.cs

```diff
@@ -47,35 +54,130 @@ public override void Initialize(AnalysisContext context)
+            var symbol = semanticModel.GetSymbolInfo(variableDeclaration.Type);
+
+            var namedTypeSymbol = symbol.Symbol as INamedTypeSymbol;
+            if (namedTypeSymbol?.TypeKind == TypeKind.Delegate)
+            {
+                var delegateParameters = namedTypeSymbol.DelegateInvokeMethod.Parameters;
+                propertiesBuilder.Add(DelegateArgumentNamesProperty, string.Join(",", delegateParameters.Select(ps => ps.Name)));
+                return true;
             }
 
-            context.ReportDiagnostic(Diagnostic.Create(Descriptor, anonymousMethod.DelegateKeyword.GetLocation()));
+            return false;
+        }
+
+        private static bool HandleAssignmentExpression(SemanticModel semanticModel, AnonymousMethodExpressionSyntax anonymousMethod, AssignmentExpressionSyntax assignmentExpressionSyntax, ImmutableDictionary<string, string>.Builder propertiesBuilder)
+        {
+            var symbol = semanticModel.GetSymbolInfo(assignmentExpressionSyntax.Left);
+
+            var eventSymbol = symbol.Symbol as IEventSymbol;
+            if (eventSymbol?.Type.TypeKind == TypeKind.Delegate)
+            {
+                var delegateParameters = ((INamedTypeSymbol)eventSymbol.Type).DelegateInvokeMethod.Parameters;
+                propertiesBuilder.Add(DelegateArgumentNamesProperty, string.Join(",", delegateParameters.Select(ps => ps.Name)));
+                return true;
+            }
+
+            return false;
+        }
+
+        private static ParameterListSyntax GetDelegateParameterList(SymbolInfo originalSymbolInfo, int argumentIndex)
+        {
+            // Determine the parameter list from the method that is invoked, as delegates without parameters are allowed, but they cannot be replaced by a lambda without parameters.
+            var methodSymbol = (IMethodSymbol)originalSymbolInfo.Symbol;
+            var delegateType = (INamedTypeSymbol)methodSymbol.Parameters[argumentIndex].Type;
+            var delegateParameters = delegateType.DelegateInvokeMethod.Parameters;
+
+            var syntaxParameters = GetSyntaxParametersFromSymbolParameters(delegateParameters);
+
+            return SyntaxFactory.ParameterList(SyntaxFactory.SeparatedList(syntaxParameters));
+        }
+
+        private static ImmutableArray<ParameterSyntax> GetSyntaxParametersFromSymbolParameters(ImmutableArray<IParameterSymbol> symbolParameters)
+        {
+            var result = ImmutableArray.CreateBuilder<ParameterSyntax>(symbolParameters.Length);
+
+            foreach (var symbolParameter in symbolParameters)
+            {
+                var syntaxParameter = GetParameterSyntaxFromParameterSymbol(symbolParameter);
+                result.Add(syntaxParameter);
+            }
+
+            return result.ToImmutable();
+        }
+
+        private static ParameterSyntax GetParameterSyntaxFromParameterSymbol(IParameterSymbol symbolParameter)
```

Comment:
❓ Is it possible to move this logic into the code fix instead? Moving it would improve the performance of the main analyzer by not requiring as much symbol analysis, and then the code fix is a bit slower but at that point users have explicitly requested the system take action.

If the new analysis logic affects when/where diagnostics are reported (and thus can't be moved), can you call those cases out?

Reply:
I'll have a look

Reply:
@sharwell I've moved the code necessary to determine the correct parameter list to the code fix. Can you have a look again?